File: Programming Fundamentals/C#/Fundamentals/09.Regular Expressions/[MoreExercises]/P01_WinningTicket/P01_WinningTicket.cs
namespace P01_WinningTicket
{
    using System;
    using System.Linq;
    using System.Text.RegularExpressions;

    class P01_WinningTicket
    {
        static void Main(string[] args)
        {
            string[] tickets = Console.ReadLine()
                                      .Split(new[] { ' ', ',' }, StringSplitOptions.RemoveEmptyEntries)
                                      .Select(x => x.Trim())
                                      .ToArray();

            string pattern = @"(\@{6,}|\${6,}|\^{6,}|\#{6,})";

            foreach (string ticket in tickets)
            {
                if (ticket.Length != 20)
                {
                    Console.WriteLine("invalid ticket");
                    continue;
                }

                Match leftSide = Regex.Match(ticket.Substring(0, 10), pattern);
                Match rightSide = Regex.Match(ticket.Substring(10), pattern);

                if (!leftSide.Success || !rightSide.Success)
                {
                    Console.WriteLine($"ticket \"{ticket}\" - no match");
                    continue;
                }

                int minLen = Math.Min(leftSide.Length, rightSide.Length);
                string left = leftSide.Value.Substring(0, minLen);
                string right = rightSide.Value.Substring(0, minLen);
                if (left.Equals(right))
                {
                    Console.WriteLine(left.Length == 10 ? $"ticket \"{ticket}\" - {minLen}{left.Substring(0, 1)} Jackpot!" :
                                                          $"ticket \"{ticket}\" - {minLen}{left.Substring(0, 1)}");
                }
                else
                {
                    Console.WriteLine($"ticket \"{ticket}\" - no match");
                }
            }
        }
    }
}

File: Programming Fundamentals/JS/09.Regular Expressions/[MoreExercises]/p01_WinningTicket.js
function solve(tickets = []) {
  tickets = tickets[0].split(/[, ]+/g);

  let pattern = /(\@{6,}|\${6,}|\^{6,}|\#{6,})/g;

  if (tickets !== null) {
    for (let ticket of tickets) {
      if (ticket.length !== 20) {
        console.log('invalid ticket');
        continue;
      }

      let leftSide = ticket.substring(0, 10).match(pattern);
      let rightSide = ticket.substring(10).match(pattern);

      if (leftSide === null || rightSide === null) {
        console.log(`ticket "${ticket}" - no match`);
        continue;
      }

      let minLen = Math.min(leftSide[0].length, rightSide[0].length);
      let left = leftSide[0].substring(0, minLen);
      let right = rightSide[0].substring(0, minLen);

      if (left === right) {
        console.log(left.length === 10 ? `ticket "${ticket}" - ${minLen}${left.substring(0, 1)} Jackpot!` :
                                         `ticket "${ticket}" - ${minLen}${left.substring(0, 1)}`);
      } else {
        console.log(`ticket "${ticket}" - no match`);
      }
    }
  }
}

// Don't copy the calling of the function in judge, you won't get any points, just the code above
solve(['Cash$$$$$$Ca$$$$$$sh']);
solve(['$$$$$$$$$$$$$$$$$$$$, aabb  , th@@@@@@eemo@@@@@@ey']);
solve(['validticketnomatch:(']);
